numba/_dispatcher.c

```diff
@@ -325,17 +328,21 @@ call_cfunc(DispatcherObject *self, PyObject *cfunc, PyObject *args, PyObject *kw
         if (PyDict_SetItemString(globals, "__builtins__", builtins)) {
             goto error;
         }
-        frame = PyFrame_New(tstate, code, globals, NULL);
+
+        store_CO_OPTIMIZED = code->co_flags & CO_OPTIMIZED;
+        /* unset the CO_OPTIMIZED flag, make the frame get a new locals dict */
+        code->co_flags &= 0xFFFE;
+
+        frame = PyFrame_New(tstate, code, globals, locals);
         if (frame == NULL) {
             goto error;
         }
         /* Populate the 'fast locals' in `frame` */
-        Py_XDECREF(frame->f_locals);
-        frame->f_locals = locals;
-        Py_XINCREF(frame->f_locals);
         PyFrame_LocalsToFast(frame, 0);
         tstate->frame = frame;
         C_TRACE(result, fn(PyCFunction_GET_SELF(cfunc), args, kws));
+        /* write changes back to locals? */
+        PyFrame_FastToLocals(frame);
```

Comment:
I think this does something more than the manual attempt at resetting https://github.com/numba/numba/pull/3638/files#diff-4a1cc89a13f0d44d41dfa7e993591bffR336-R337 which might be why this seems to work?